File: feedSystem_video_python/frontend/src/api/feed.ts
import { postJson } from './client'
import type { ListByFollowingResponse, ListByTagResponse, ListLatestResponse, ListLikesCountResponse } from './types'

export function listLatest(limit: number, latestTime = 0): Promise<ListLatestResponse> {
  /** 查询最新 Feed；latest_time 为上一页最后一条视频的毫秒时间戳。 */
  return postJson<ListLatestResponse>('/feed/listLatest', { limit, latest_time: latestTime })
}

export function listLikesCount(
  limit: number,
  likesCountBefore?: number | null,
  idBefore?: number | null,
): Promise<ListLikesCountResponse> {
  /** 查询点赞榜 Feed；使用 likes_count + id 复合游标。 */
  return postJson<ListLikesCountResponse>('/feed/listLikesCount', {
    limit,
    likes_count_before: likesCountBefore ?? null,
    id_before: idBefore ?? null,
  })
}

export function listByFollowing(limit: number, latestTime = 0): Promise<ListByFollowingResponse> {
  /** 查询关注 Feed；需要登录，latest_time 为上一页最后一条视频的毫秒时间戳。 */
  return postJson<ListByFollowingResponse>('/feed/listByFollowing', { limit, latest_time: latestTime })
}

export function listByTag(tagName: string, limit: number, latestTime = 0): Promise<ListByTagResponse> {
  /** 查询标签 Feed；tag_name 可以传 Python 或 #Python，后端会统一处理。 */
  return postJson<ListByTagResponse>('/feed/listByTag', {
    tag_name: tagName,
    limit,
    latest_time: latestTime,
  })
}

File: feedSystem_video_python/backend/app/api/feed.py
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.auth import CurrentUser, get_current_user, get_optional_user
from app.database import get_db
from app.repositories.feed_repo import FeedRepository
from app.repositories.like_repo import LikeRepository
from app.schemas.feed import (
    ListLatestRequest,
    ListLatestResponse,
    ListByFollowingRequest,
    ListByFollowingResponse,
    ListByTagRequest,
    ListByTagResponse,
    ListLikesCountRequest,
    ListLikesCountResponse,
)
from app.services.feed_service import FeedService


router = APIRouter(prefix="/feed", tags=["feed"])


def get_feed_service(db: AsyncSession = Depends(get_db)) -> FeedService:
    """FastAPI 依赖函数：组装 FeedService，复用 Feed 和 Like 查询能力。"""
    return FeedService(FeedRepository(db), LikeRepository(db))


@router.post("/listLatest", response_model=ListLatestResponse)
async def list_latest(
    req: ListLatestRequest,
    viewer: CurrentUser | None = Depends(get_optional_user),
    service: FeedService = Depends(get_feed_service),
) -> ListLatestResponse:
    """最新 Feed 接口：软鉴权，使用 create_time 毫秒游标分页。"""
    return await service.list_latest(req.limit, req.latest_time, viewer)


@router.post("/listLikesCount", response_model=ListLikesCountResponse)
async def list_likes_count(
    req: ListLikesCountRequest,
    viewer: CurrentUser | None = Depends(get_optional_user),
    service: FeedService = Depends(get_feed_service),
) -> ListLikesCountResponse:
    """点赞榜 Feed 接口：软鉴权，使用 likes_count + id 复合游标分页。"""
    return await service.list_likes_count(
        limit=req.limit,
        likes_count_before=req.likes_count_before,
        id_before=req.id_before,
        viewer=viewer,
    )


@router.post("/listByFollowing", response_model=ListByFollowingResponse)
async def list_by_following(
    req: ListByFollowingRequest,
    viewer: CurrentUser = Depends(get_current_user),
    service: FeedService = Depends(get_feed_service),
) -> ListByFollowingResponse:
    """关注 Feed 接口：必须登录，只看自己关注作者发布的视频。"""
    return await service.list_by_following(req.limit, req.latest_time, viewer)


@router.post("/listByTag", response_model=ListByTagResponse)
async def list_by_tag(
    req: ListByTagRequest,
    viewer: CurrentUser | None = Depends(get_optional_user),
    service: FeedService = Depends(get_feed_service),
) -> ListByTagResponse:
    """标签 Feed 接口：软鉴权，按 #tag 查询视频列表。"""
    return await service.list_by_tag(req.tag_name, req.limit, req.latest_time, viewer)
